htlcswitch/failure_detail.go

```diff
@@ -37,6 +41,48 @@ const (
 	// FailureDetailDownstreamHtlcAdd is returned when we fail to add a
 	// downstream htlc to our outgoing link.
 	FailureDetailDownstreamHtlcAdd
+
+	// FailureDetailCannotEncodeRoute is returned when we cannot encode the
+	// remainder of the route.
+	FailureDetailCannotEncodeRoute
+
+	// FailureDetailInvoiceAlreadyCancelled is returned when an invoice has
+	// already been cancelled when an attempt to pay is is made.
+	FailureDetailInvoiceAlreadyCancelled
+
+	// FailureDetailAmountTooLow is returned when an invoice is underpaid.
+	FailureDetailAmountTooLow
+
+	// FailureDetailExpiryTooSoon is returned when we do not accept an
+	// invoice payment because it expires too soon.
+	FailureDetailExpiryTooSoon
+
+	// FailureDetailInvoiceNotOpen is returned when an attempt is made to
+	// pay a mpp invoice which is not open.
+	FailureDetailInvoiceNotOpen
+
+	// FailureDetailAddressMismatch is returned when the payment address
+	// for a mpp invoice does not match.
+	FailureDetailAddressMismatch
+
+	// FailureDetailSetTotalMismatch is returned when the amount paid by a
+	// htlc does not match its set total.
+	FailureDetailSetTotalMismatch
+
+	// FailureDetailSetTotalTooLow is returned when a mpp set total is too
+	// low for an invoice.
+	FailureDetailSetTotalTooLow
+
+	// FailureDetailSetOverpaid is returned when a mpp set is overpaid.
+	FailureDetailSetOverpaid
+
+	// FailureDetailInvoiceNotFound is returned when an attempt is made to
+	// pay an invoice that is unknown to us.
+	FailureDetailInvoiceNotFound
+
+	// FailureDetailHodlCancel is returned when a hodl invoice is
+	// cancelled.
+	FailureDetailHodlCancel
```

Comment:
By updating `invoices` to return a `FailureDetail`? The sets don't map perfectly onto each other. Since we'd only be acting on `ResolutionResult` enums that are for failures; something like `ResultReplayToSettled` has no natural failure detail to be associated with. 

`FailureDetail` is a superset of `ResolutionResult`, so we can't do the reverse (use resolution results in htlcswitch). And I'm quite averse to trying to combine two enum types into one, too much risk of overlapping values creating bugs when one set is changed and the other is not.

Reply:
Maybe splitting the settle and fail results in two different enums in `invoices` could have made it easier to use the same (fail) enum here? But I am fine with the current mapping too.

Reply:
We did have some discussion in this direction on the [original PR](https://github.com/lightningnetwork/lnd/pull/3831#discussion_r359324071), but it's not so clean when using iota. If we end up using `ResolutionResult` for more than this, we could split out into a `ResolutionResult` interface which is implemented by `Settle` and `Fail` structs that contain the the enum + extra info such as preimage but I think that's overkill for now.

Reply:
Thanks for refreshing that argument. Yes, the abstract result interface could be nice to make it a bit stricter and prevent the mapping. But can indeed be done at any time later too.